File: tests/algorithms/x/Java/ciphers/atbash.java
public class Main {

    static int index_of(String s, String c) {
        int i = 0;
        while (i < _runeLen(s)) {
            if ((s.substring(i, i+1).equals(c))) {
                return i;
            }
            i = i + 1;
        }
        return (-1);
    }

    static String atbash(String sequence) {
        String lower = "abcdefghijklmnopqrstuvwxyz";
        String upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        String lower_rev = "zyxwvutsrqponmlkjihgfedcba";
        String upper_rev = "ZYXWVUTSRQPONMLKJIHGFEDCBA";
        String result = "";
        int i_1 = 0;
        while (i_1 < _runeLen(sequence)) {
            String ch = sequence.substring(i_1, i_1+1);
            int idx = index_of(lower, ch);
            if (idx != (-1)) {
                result = result + lower_rev.substring(idx, idx+1);
            } else {
                int idx2 = index_of(upper, ch);
                if (idx2 != (-1)) {
                    result = result + upper_rev.substring(idx2, idx2+1);
                } else {
                    result = result + ch;
                }
            }
            i_1 = i_1 + 1;
        }
        return result;
    }
    public static void main(String[] args) {
        System.out.println(atbash("ABCDEFGH"));
        System.out.println(atbash("123GGjj"));
        System.out.println(atbash("testStringtest"));
        System.out.println(atbash("with space"));
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }
}

File: tests/algorithms/x/Kotlin/ciphers/atbash.kt
var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

fun index_of(s: String, c: String): Int {
    var i: Int = 0
    while (i < s.length) {
        if (s[i].toString() == c) {
            return i
        }
        i = i + 1
    }
    return 0 - 1
}

fun atbash(sequence: String): String {
    var lower: String = "abcdefghijklmnopqrstuvwxyz"
    var upper: String = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    var lower_rev: String = "zyxwvutsrqponmlkjihgfedcba"
    var upper_rev: String = "ZYXWVUTSRQPONMLKJIHGFEDCBA"
    var result: String = ""
    var i: Int = 0
    while (i < sequence.length) {
        var ch: String = sequence[i].toString()
        var idx: Int = index_of(lower, ch)
        if (idx != (0 - 1)) {
            result = result + lower_rev[idx].toString()
        } else {
            var idx2: Int = index_of(upper, ch)
            if (idx2 != (0 - 1)) {
                result = result + upper_rev[idx2].toString()
            } else {
                result = result + ch
            }
        }
        i = i + 1
    }
    return result
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        println(atbash("ABCDEFGH"))
        println(atbash("123GGjj"))
        println(atbash("testStringtest"))
        println(atbash("with space"))
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
